[DataGrid] Add methods to import and export the state

flaviendelangle opened the pull request:
Closes #820

[Doc preview](https://deploy-preview-3593--material-ui-x.netlify.app/components/data-grid/state/#save-and-restore-the-state)

## Public API

```tsx
// Export all the exportable state
const state = apiRef.current.exportState();

// Import all the exportable state
apiRef.current.restoreState(state);

// Or import some parts of the exportable state
apiRef.current.restoreState({ pinning: state.pinning });

// Or pass the exportable state to the `prop.initialState`
<DataGrid initialState={state} />
```

## Internals

Each hook with an initial state must also register two pre-processors:

- `exportState` which takes a `GridInitialState` and add the initial state from this hook to it

```ts
  const stateExportPreProcessing = React.useCallback<
    GridPreProcessor<GridPreProcessingGroup.exportState>
  >(
    (prevState) => {
      return {
        ...prevState,
        pagination: {
          ...prevState.pagination,
          pageSize: gridPageSizeSelector(apiRef.current.state),
        },
      };
    },
    [apiRef],
  );
```

- `restoreState` which takes the `GridInitialState` to restore and updates the state accordingly.

```ts
  const stateRestorePreProcessing = React.useCallback<
    GridPreProcessor<GridPreProcessingGroup.restoreState>
  >(
    (params, context) => {
      const pageSize = context.stateToRestore.pagination?.pageSize;
      if (pageSize != null) {
        apiRef.current.setState(setStatePageSize(pageSize));
      }
      return params;
    },
    [apiRef],
  );
```

  This method can also return a callback that will be ran after all the initial states are restored. This can be usefull to apply the sorting / filtering for instance.

I don't use directly `setFilterModel`, `setPageSize`, etc... to avoid multiple `forceUpdates` which causes a lot of re-renders.

m4theushw commented on docs/src/pages/components/data-grid/state/RestoreStateApiRef.tsx:
```diff
@@ -0,0 +1,158 @@
+      </Stack>
+      <Stack direction="row" spacing={1}>
+        {Object.entries(state.views).map(([viewId, view]) => (
+          <Chip
```
You could use [Tabs](https://mui.com/components/tabs) or a Select to load a previous session. The Chip doesn't have a selected state.

flaviendelangle replied:
I did a basic implementation based on the Qonto example

m4theushw reviewed (commented):
Despite the title of #820, it asks for being able to save the order of the columns, we're not allowing this here yet.

We could export only those sub-states which are different from their initial state. Currently, `pinnedColumns` is being exported even if no column is pinned. If one day we change the initial value, users don't need to care about normalizing saved states in a database.

flaviendelangle commented on packages/grid/x-data-grid-pro/src/tests/statePersistence.DataGridPro.test.tsx:
```diff
@@ -0,0 +1,238 @@
+      expect(getColumnValues(0)).to.deep.equal(['Titanic', 'Star Wars: The Force Awakens']);
+    });
+
+    // Not sure how to make that one work, if the `pageSize` update is async, we try to update the `page` with an outdated `pageSize` which can cause problem.
```
These two props are very entangled.
Defined them together would indeed have some benefits in v6.
That's worth considering.

m4theushw replied:
If we agreed that this is a limitation, the commented code could be removed.

flaviendelangle commented on docs/src/pages/components/data-grid/state/RestoreStateInitialState.tsx:
```diff
@@ -0,0 +1,93 @@
+  });
+
+  if (loading) {
+    return null;
```
https://github.com/mui-org/material-ui-x/pull/3593/files#diff-fdb681e484a1f446b405b88cac0fe5a5e7e97d0444ba1f89679fa09e15da8c19R64

It works for everything except the pagination
We could make the behavior evolve in v6 (or before if we consider this behavior change non-breaking)
This is slightly related to https://github.com/mui-org/material-ui-x/issues/3602 and https://github.com/mui-org/material-ui-x/issues/1685, our pagination constraint strategy can be limiting right now.

flaviendelangle commented on packages/grid/_modules_/grid/hooks/features/preferencesPanel/useGridPreferencesPanel.ts:
```diff
@@ -59,6 +71,48 @@ export const useGridPreferencesPanel = (
+  /**
+   * PRE-PROCESSING
+   */
+  const stateExportPreProcessing = React.useCallback<
```
To be fair, the preference panel initial state is here because we are using it a lot in tests and it does not hurt for users.
But it's not a very interesting thing to export / restore I agree.

flaviendelangle commented:
> We could export only those sub-states which are different from their initial state. Currently, `pinnedColumns` is being exported even if no column is pinned. If one day we change the initial value, users don't need to care about normalizing saved states in a database. I guess a `isDeepEqual`, for each key, between the current value and the default one does the trick.

I can add a small check in each pre-processor to not include its value if it's not interesting to do it.

joserodolfofreitas reviewed (approved):
It's looking good. I left just a couple of suggestions for the docs.

flaviendelangle commented on docs/src/pages/components/data-grid/state/state.md:
```diff
@@ -55,9 +56,36 @@ Some selectors are yet to be documented.
+The current state of the grid can be exported using `apiRef.current.exportState()`.
+It can then be restored by either passing it to the `initialState` prop or to the `apiRef.current.restoreState()` method.
+
+If one of the sub-state is being controlled, the grid will call the callback linked to that model (`onFilterModelChange` if `filterModel` is being controlled for instance).
```
`on RestoreState` ?

joserodolfofreitas replied:
Yeah, that is misleading. I updated the suggestion.

flaviendelangle commented on packages/grid/_modules_/grid/hooks/features/statePersistence/useGridStatePersistence.ts:
```diff
@@ -0,0 +1,44 @@
+        callback();
+      });
+
+      apiRef.current.forceUpdate();
```
Right now all callbacks do re-render so I guess we can just add a 

```suggestion
      if (response.callbacks.length === 0) {
        apiRef.current.forceUpdate();
      }
```

flaviendelangle commented:
> Despite the title of #820, it asks for being able to save the order of the columns, we're not allowing this here yet. I think initially the idea was to export the entire state as it is, then we discovered that without cleaning it that wouldn't be possible. Now #820 seems more of an umbrella issue.


I have #3448 for topics around adding new sub-states.
I will rename it once we have the import / export

benjimac93 commented:
> Despite the title of #820, it asks for being able to save the order of the columns, we're not allowing this here yet. I think initially the idea was to [export the entire state](https://github.com/mui-org/material-ui-x/issues/820#issuecomment-756962564) as it is, then we discovered that without cleaning it that wouldn't be possible. Now #820 seems more of an umbrella issue.
> 
> We could export only those sub-states which are different from their initial state. Currently, `pinnedColumns` is being exported even if no column is pinned. If one day we change the initial value, users don't need to care about normalizing saved states in a database. I guess a `isDeepEqual`, for each key, between the current value and the default one does the trick.

Is there a plan to add support for the column orders in a subsequent release? For us, this would be one of the most important parts of state to save/restore

flaviendelangle commented:
> Is there a plan to add support for the column orders in a subsequent release? For us, this would be one of the most important parts of state to save/restore

No, we did not start working on it.
And as it is independent in our code from the save/restore process itself, we have interest in splitting the work to release a first version of the save/restore as soon as possible to have feedbacks.

tielushko commented:
I support the @benjimac93 notion of having the order of columns saved as the part of initial state as well as the width of the individual column. It would be a very useful feature for our end users.

benjimac93 commented:
> > Is there a plan to add support for the column orders in a subsequent release? For us, this would be one of the most important parts of state to save/restore
> 
> No, we did not start working on it. And as it is independent in our code from the save/restore process itself, we have interest in splitting the work to release a first version of the save/restore as soon as possible to have feedbacks.

Understood it's not in scope for the initial release of this feature. Is there a rough timeline for when it will be in scope in a future release though? Trying to decide if we use MUI Pro or Premium for our product rebuild or if we need to explore other options

flaviendelangle commented:
@benjimac93 we could probably try to release this feature by the end of Q1 as it is not a major peace of work.
But can't guarantee anything.

benjimac93 commented:
> @benjimac93 we could probably try to release this feature by the end of Q1 as it is not a major peace of work. But can't guarantee anything.

I can work with that. Component is awesome as an aside !

m4theushw commented on docs/src/pages/components/data-grid/state/RestoreStateInitialState.tsx:
```diff
@@ -0,0 +1,99 @@
+      <Box sx={{ width: '100%', height: 400, bgcolor: 'background.paper' }}>
+        <WrappedDataGridPro unMount={unMountGrid} initialState={savedState} />
+        {!!savedState && (
+          <Alert severity="info" style={{ marginBottom: 8 }}>
```
The `Alert` needs to be outside the fixed-height box.

![image](https://user-images.githubusercontent.com/42154031/151802406-e4748a08-ec20-46c9-98fc-1ae424e168c8.png)

m4theushw commented on docs/src/pages/components/data-grid/state/RestoreStateApiRef.tsx:
```diff
@@ -0,0 +1,334 @@
+                      onSelect={handleSetActiveView}
+                    />
+                  ))}
+                  <Divider />
```
Something is odd with this demo. Having the text input inside the list doesn't look very good. What about putting the input outside the menu, like in a toolbar?

![image](https://user-images.githubusercontent.com/42154031/151804977-cb32b525-c904-4d95-9127-f75627bb53aa.png)

Because of the way this variant works it appears misaligned when not fucused.

![image](https://user-images.githubusercontent.com/42154031/151805239-10b1546c-5421-49c1-89b2-ddcb20fa4d8a.png)

flaviendelangle replied:
I did not finalize this demo, as I wanted feedbacks about the global idea before spending to much times on the details.
But I agree that it is not 100% clean right now :+1:

m4theushw reviewed (commented):
> I can add a small check in each pre-processor to not include its value if it's not interesting to do it.

Will this be done in this PR?

flaviendelangle commented:
> Will this be done in this PR?

Yes I think it's better to do it, I will add it today or tomorrow

flaviendelangle commented:
@m4theushw I added the default value stripping
And modified the pinned columns state init to handle the control model synchronously and storing the initial model given by the user as provided in state.

m4theushw reviewed (approved):
The demo is not perfect, but I'm approving for the API which is good. 

<img width="300" src="https://user-images.githubusercontent.com/42154031/152078850-256f7f76-a324-4258-bfab-6e358e1e8fe5.png" alt="image" style="max-width: 100%;">

Looking at the feedback received, what users are really asking is to restore the order of the columns. What about keeping the docs commented (demos included) until we add support for this part?

flaviendelangle commented:
@m4theushw I can merge this one without the doc visible and start working on the columns size and order yes.

gminova commented:
@flaviendelangle Great work and so happy this is merged now! I was wondering why is the scrollPosition not part of the state? and is there a way to set an initialScrollPosition for the grid just like there is an initialState prop?

flaviendelangle commented:
The scroll position is not on our internal state at all because it is too volatile.
Each state update is supposed to cause a re-render and we don't want to re-render on scroll.

What is your use-case exactly ? Would you like to scroll to a position in pixel or to a given row on mount ?
